File: src/main/java/tech/zolhungaj/amqapi/servercommands/objects/SongInfo.java
package tech.zolhungaj.amqapi.servercommands.objects;

import java.util.List;
import java.util.Map;
import java.util.Optional;

import com.squareup.moshi.Json;

public record SongInfo(
        @Json(name = "songName")
        String songName,
        @Json(name = "animeGenre")
        List<String> animeGenre,

        @Json(name = "animeNames")
        MainAnimeNames mainAnimeNames,
        @Json(name = "artist")
        String artist,

        @Json(name = "animeDifficulty")
        Optional<Double> animeDifficulty,

        @Json(name = "altAnimeNames")
        List<String> alternativeAnimeNames,

        @Json(name = "animeScore")
        Optional<Double> animeScore,
        @Json(name = "type")
        SongType type,

        @Json(name = "urlMap")
        Map<String, Map<String, String>> urlMap,

        @Json(name = "typeNumber")
        int typeNumber,
        @Json(name = "annId")
        int annId,

        @Json(name = "vintage")
        String vintage,
        @Json(name = "animeTags")
        List<String> animeTags,

        @Json(name = "siteIds")
        AnimeListSiteShowIds siteIds,
        @Json(name = "animeType")
        String animeType,
        @Json(name = "highRisk")
        Boolean highRisk,

        @Json(name = "altAnimeNamesAnswers")
        List<String> alternativeAnimeNamesAnswers)
{}

File: src/main/java/tech/zolhungaj/amqapi/servercommands/objects/SongInfo.kt
package tech.zolhungaj.amqapi.servercommands.objects

import com.squareup.moshi.Json

@JvmRecord
data class SongInfo(
    @Json(name = "songName") val songName: String,
    @Json(name = "animeGenre") val animeGenre: List<String>,

    @Json(name = "animeNames") val mainAnimeNames: MainAnimeNames,
    @Json(name = "artist") val artist: String,

    @Json(name = "animeDifficulty") val animeDifficulty: AnimeDifficulty,
    @Json(name = "seasonInfo") val season: SeasonInfo?,

    @Json(name = "altAnimeNames") val alternativeAnimeNames: List<String>,

    @Json(name = "animeScore") val animeScore: Double?,
    @Json(name = "composerInfo") val composerInfo: ArtistInfo?,

    @Json(name = "type") val type: SongType,
    @Json(name = "dub") val isDub: Boolean,
    @Json(name = "artistInfo") val artistInfo: ArtistInfo?,


    @Json(name = "videoTargetMap") val videoTargetMap: Map<String, Map<String, String>>,

    @Json(name = "typeNumber") val typeNumber: Int,
    @Json(name = "annId") val annId: Int,

    @Json(name = "vintage") val vintage: Vintage,
    @Json(name = "animeTags") val animeTags: List<String>,

    @Json(name = "siteIds") val siteIds: AnimeListSiteShowIds,
    @Json(name = "arrangerInfo") val arrangerInfo: ArtistInfo?,

    @Json(name = "animeType") val animeType: String,
    @Json(name = "highRisk") val isHighRisk: Boolean,
    @Json(name = "rebroadcast") val isRebroadcast: Boolean,

    @Json(name = "altAnimeNamesAnswers") val alternativeAnimeNamesAnswers: List<String>
){
    data class SeasonInfo(val number: String?, val name: String)
    data class ArtistInfo(val name: String, val artistId: Int?, val groupId: Int?)
    @JvmRecord
    data class Vintage(val key: String, val data: VintageData){
        data class VintageData(val year: Int)
    }
}